WIP: Make links more accessible.

rassie opened the pull request:
This is a work-in-progress PR which intends to fix #1709. 

The first iteration of this PR only includes changes to `<CloneButton />`, while still adding all necessary machinery for other buttons. It is meant for a first discussion on implementation and code style before applying the same style to the other buttons.

**Problems**
I do not like the way `react-router` recommends testing `location` objects with `render` property (which is what this patch is based on). It feels unnecessarily verbose und complex, but I have yet to see a method that works at all and looks better. On the positive side, this kind of testing actually tests the result instead of implementation -- previously the test checked whether a function has been called, while not verifying that the end result is what is expected.

**Work to do**
A lot of other components need to be updated apart from the buttons:

- [ ] Sider
- [ ] SimpleTable
- [ ] etc.

omeraplak commented on packages/antd/src/components/buttons/clone/index.tsx:
```diff
@@ -67,18 +69,24 @@ export const CloneButton: React.FC<CloneButtonProps> = ({
-            title={createButtonDisabledTitle()}
-            {...rest}
+        <Link
+            to={cloneUrl(propResourceName ?? resource.route!, id!)}
```
Should we get property for `<Link>` component? like a ButtonProps

rassie replied:
Have to lookup the workings of `ButtonProps` first, will get back to you.

`Link` component however is problematic in itself -- I think we might need a wrapper since otherwise `RouterProvider` is not being abstracted away, but instead you have to be aware whether it's `react-router` or `react-router-v6` or something else completely.

omeraplak replied:
Yes, I agree, this is a problem we have to overcome. Would a `LinkProps` with type `any` be too confusing?

omeraplak replied:
Great, we trust you and your friends!! 🥇

rassie commented on packages/antd/src/hooks/table/useTable/useTable.ts:
```diff
@@ -15,12 +15,17 @@ import {
     mapAntdFilterToCrudFilter,
 } from "../../../definitions/table";
+import { PaginationLink } from "./paginationLink";
+import { createElement } from "react";
```
Consider it done!

omeraplak commented on packages/antd/src/hooks/table/useTable/useTable.ts:
```diff
@@ -114,6 +119,11 @@ export const useTable = <
         form: form,
     });
 
+    const { syncWithLocation: syncWithLocationContext } = useSyncWithLocation();
```
If we add a function like `createLinkForSyncWithLocation` to the useTable hook in the core.

```
const URL = createLinkForSyncWithLocation({
                         pagination: {
                             pageSize,
                             current: page,
                         },
                         sorter,
                         filters,
                     });
```
So we can use it in all table packages. now Ant Design and react-table. In the future in our Material UI package.

omeraplak replied:
Wow so good! Congrats!

rassie commented:
@omeraplak I'm updating the tests after recent move on `master` from `react-router` to `react-router-v6`. It certainly feels like the router abstraction is not working on test level, since it's basically testing integration with `react-router-v6` and not with the router provider. But I suppose it's a greater problem, so I'm just adapting to the existing code for now.

omeraplak commented:
> @omeraplak I'm updating the tests after recent move on `master` from `react-router` to `react-router-v6`. It certainly feels like the router abstraction is not working on test level, since it's basically testing integration with `react-router-v6` and not with the router provider. But I suppose it's a greater problem, so I'm just adapting to the existing code for now.

I'm sorry for the confusion. I also agree with you

rassie commented:
@omeraplak I've updated the tests to work with `react-router` v6 and moved the `TestWrapper` to use `BrowserRouter` while we at it, since I have no other option to have `location.pathname` checked in tests. If all the tests complete (which I'm not quite sure of), I'd consider this to be ready to go, but I'd be happy to incorporate further feedback if you have any.

omeraplak commented:
> @omeraplak I've updated the tests to work with `react-router` v6 and moved the `TestWrapper` to use `BrowserRouter` while we at it, since I have no other option to have `location.pathname` checked in tests. If all the tests complete (which I'm not quite sure of), I'd consider this to be ready to go, but I'd be happy to incorporate further feedback if you have any.

Hey @rassie ,
`BrowserRouter` looks cool. Thank you! Finally, https://github.com/pankod/refine/pull/1710/files#r848168971 Did you have a chance to look at this comment? If you don't have time, we can handle it. Thanks!

rassie commented:
Yes, indeed, I've looked at it. Let me push some changes :)

omeraplak commented:
I will merge it soon but we need to update some documents before we release it. Thank you very much your amazing effort! @rassie 🚀
